synapse/handlers/user_directory.py

```diff
@@ -229,25 +229,22 @@ async def _handle_deltas(self, deltas: List[Dict[str, Any]]) -> None:
                 ) or await self.store.should_include_local_user_in_dir(state_key)
                 if include_in_dir:
                     if change is MatchChange.no_change:
-                        # Handle any profile changes
-                        await self._handle_profile_change(
-                            state_key, room_id, prev_event_id, event_id
-                        )
+                        # Handle any profile changes for remote users.
+                        # (For local users we are not forced to scan membership
+                        # events; instead the rest of the application calls
+                        # `_handle_local_profile_change`.)
+                        if not self.is_mine_id(state_key):
+                            await self._handle_profile_change(
+                                state_key, room_id, prev_event_id, event_id
+                            )
                         continue
 
                     if change is MatchChange.now_true:  # The user joined
-                        event = await self.store.get_event(event_id, allow_none=True)
-                        # It isn't expected for this event to not exist, but we
-                        # don't want the entire background process to break.
-                        if event is None:
-                            continue
-
-                        profile = ProfileInfo(
-                            avatar_url=event.content.get("avatar_url"),
-                            display_name=event.content.get("displayname"),
-                        )
-
-                        await self._handle_new_user(room_id, state_key, profile)
+                        if not self.is_mine_id(state_key):
+                            await self._upsert_directory_entry_for_remote_user(
```

Comment:
I haven't really looked at it, but I wonder if this could also re-use `_handle_profile_change`?